File: 1277. 统计全为 1 的正方形子矩阵.rs
impl Solution {
    pub fn count_squares(matrix: Vec<Vec<i32>>) -> i32 {
        use std::cmp::min;
        let mut dp: Vec<Vec<i32>> = vec![vec![0; matrix[0].len()]; matrix.len()];
        let mut ans: i32 = 0;
        for i in 0..matrix.len() {
            for j in 0..matrix[0].len() {
                if i == 0 || j == 0 {
                    dp[i][j] = matrix[i][j];
                } else if matrix[i][j] == 0 {
                    dp[i][j] = 0;
                } else {
                    dp[i][j] = min(dp[i][j - 1], min(dp[i - 1][j], dp[i - 1][j - 1])) + 1;
                }
                ans += dp[i][j];
            }
        }
        ans
    }
}

File: leetcode/1277. 统计全为 1 的正方形子矩阵.go
func countSquares(matrix [][]int) int {
	m := len(matrix)
	n := len(matrix[0])
	dp := make([][]int, m)
	ans := 0
	for i := range dp {
		dp[i] = make([]int, n)
	}
	for i := 0; i < m; i++ {
		for j := 0; j < n; j++ {
			if i == 0 || j == 0 {
				dp[i][j] = matrix[i][j]
			} else if matrix[i][j] == 0 {
				dp[i][j] = 0
			} else {
				dp[i][j] = Min(dp[i][j-1], dp[i-1][j], dp[i-1][j-1]) + 1
			}
			ans += dp[i][j]
		}
	}
	return ans
}

func Min(a, b, c int) int {
	if a <= b && a <= c {
		return a
	} else if b <= a && b <= c {
		return b
	} else if c <= a && c <= b {
		return c
	}
	return 0
}